pkg/server/server_controller.go

```diff
@@ -172,20 +175,146 @@ func newServerController(
+// the running servers accordingly.
+func (c *serverController) start(ctx context.Context, ie isql.Executor) error {
+	// We perform one round of updates synchronously, to ensure that
+	// any tenants already in service mode SHARED get a chance to boot
+	// up before we signal readiness.
+	if err := c.startInitialSecondaryTenantServers(ctx, ie); err != nil {
+		return err
+	}
+
+	// Then update them async.
+	return c.stopper.RunAsyncTask(ctx, "update-tenant-services", func(ctx context.Context) {
+		const watchInterval = time.Second
+		for {
+			select {
+			case <-time.After(watchInterval):
+			case <-c.stopper.ShouldQuiesce():
+				return
+			}
+
+			if err := c.updateRunningTenants(ctx, ie); err != nil {
+				log.Warningf(ctx, "cannot update running tenant services: %v", err)
+			}
+		}
+	})
+}
+
+// startInitialSecondaryTenantServers starts the servers for secondary tenants
+// that should be started during server initialization.
+func (c *serverController) startInitialSecondaryTenantServers(
+	ctx context.Context, ie isql.Executor,
+) error {
+	// The list of tenants that should have a running server.
+	reqTenants, err := c.getExpectedRunningTenants(ctx, ie)
+	if err != nil {
+		return err
+	}
+	for name := range reqTenants {
+		if name == catconstants.SystemTenantName {
+			// We already pre-initialize the entry for the system tenant.
+			continue
+		}
+		if _, err := c.createServer(ctx, name, base.TestSharedProcessTenantArgs{}); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// updateRunningTenants checks which tenants need to be started/stopped
+// and performs the necessary server lifecycle changes.
+func (c *serverController) updateRunningTenants(ctx context.Context, ie isql.Executor) error {
+	// The list of tenants that should have a running server.
+	reqTenants, err := c.getExpectedRunningTenants(ctx, ie)
+	if err != nil {
+		return err
+	}
+
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	// First check if there are any servers that shouldn't be running right now.
+	for name, srv := range c.mu.servers {
+		if !srv.shouldStop {
+			continue
+		}
+		if _, ok := reqTenants[name]; !ok {
+			// Tenant currently running, but should not be running. Shut it down.
+			if err := c.stopper.RunAsyncTask(ctx, "stop-tenant-async", func(ctx context.Context) {
+				srv.server.stop(context.Background())
+			}); err != nil {
+				// Failed to add task: shutting down.
+				return err
+			}
+		}
+	}
+
+	// Now add all the missing servers.
+	for name := range reqTenants {
+		if _, ok := c.mu.servers[name]; !ok {
+			// Server not currently running, and should be running. Start it.
+			tenantName := name // copy loop iteration variable.
+			if err := c.stopper.RunAsyncTask(ctx, "start-tenant-async", func(ctx context.Context) {
+				if _, err := c.createServer(ctx, tenantName, base.TestSharedProcessTenantArgs{}); err != nil {
+					log.Warningf(ctx, "unable to start service for tenant: %v", err)
+				}
```

Comment:
Do we want any backoff for tenants that we can't start for some reason?  Currently, we'll retry every second which seems a bit aggressive if we've gotten an error.